app/services/link_reporter_csv_service.rb

```diff
@@ -0,0 +1,66 @@
+class LinkReporterCsvService
+  def initialize(reports_dir:, organisation:)
+    @organisation = organisation
+    @reports_dir = reports_dir
+  end
+
+  def generate
+    CSV.open(file_path, "w", encoding: "UTF-8") do |csv|
+      csv << headings
+      public_editions.each { |edition| csv << row_for_edition(edition) }
+    end
+  end
+
+private
+
+  attr_reader :organisation, :reports_dir
+
+  def public_editions
+    Edition.publicly_visible.with_translations.in_organisation(organisation)
+  end
+
+  def file_path
+    reports_dir.join("#{@organisation.slug}_links_report.csv")
+  end
+
+  def headings
+    ["page", "admin link", "public timestamp", "format", "broken link count", "broken links"]
+  end
+
+  def row_for_edition(edition)
+    [
+      public_url(edition),
+      admin_url(edition),
+      timestamp(edition),
+      edition.type,
+      broken_links(edition).size,
+      broken_links(edition).join("\r\n")
+    ]
+  end
+
+  def public_url(edition)
+    Whitehall.url_maker.public_document_url(edition, host: public_host, protocol: "https")
+  end
+
+  def admin_url(edition)
+    Whitehall.url_maker.admin_edition_url(edition, host: admin_host, protocol: "https")
+  end
+
+  def timestamp(edition)
+    edition.public_timestamp.to_s
+  end
+
+  def broken_links(edition)
+    edition.link_check_reports.last.broken_links.map(&:uri)
```

Comment:
We might have to put in a check here for if the edition has no link check reports? Not sure whether the best thing to do is to include it in the CSV file with a note or just not include it...

Reply:
👍 good spot, I've stuck with the existing functionality of not including the edition in the CSV if there is no report. This is only really to handle very edge cases because as the worker task will be running every night and so it will be very rare that an edition doesn't have a link check report.